packages/collaboration/src/browser/collaboration-file-system-provider.ts

```diff
@@ -0,0 +1,119 @@
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import * as Y from 'yjs';
+import { Disposable, Emitter, Event, URI } from '@theia/core';
+import { injectable } from '@theia/core/shared/inversify';
+import {
+    FileChange, FileDeleteOptions,
+    FileOverwriteOptions, FileSystemProviderCapabilities, FileType, Stat, WatchOptions, FileSystemProviderWithFileReadWriteCapability, FileWriteOptions
+} from '@theia/filesystem/lib/common/files';
+import { ProtocolBroadcastConnection, Workspace, Peer } from 'open-collaboration-protocol';
+
+export namespace CollaborationURI {
+
+    export const scheme = 'collaboration';
+
+    export function create(workspace: Workspace, path?: string): URI {
+        return new URI(`${scheme}:///${workspace.name}${path ? '/' + path : ''}`);
+    }
+}
+
+@injectable()
+export class CollaborationFileSystemProvider implements FileSystemProviderWithFileReadWriteCapability {
+
+    capabilities = FileSystemProviderCapabilities.FileReadWrite;
+
+    protected _readonly: boolean;
+
+    get readonly(): boolean {
+        return this._readonly;
+    }
+
+    set readonly(value: boolean) {
+        if (this._readonly !== value) {
+            this._readonly = value;
+            if (value) {
+                this.capabilities |= FileSystemProviderCapabilities.Readonly;
+            } else {
+                this.capabilities &= ~FileSystemProviderCapabilities.Readonly;
+            }
+            this.onDidChangeCapabilitiesEmitter.fire();
+        }
+    }
+
+    constructor(readonly connection: ProtocolBroadcastConnection, readonly host: Peer, readonly yjs: Y.Doc) {
+    }
+
+    protected encoder = new TextEncoder();
+    protected decoder = new TextDecoder();
+    protected onDidChangeCapabilitiesEmitter = new Emitter<void>();
+    protected onDidChangeFileEmitter = new Emitter<readonly FileChange[]>();
+    protected onFileWatchErrorEmitter = new Emitter<void>();
+
+    get onDidChangeCapabilities(): Event<void> {
+        return this.onDidChangeCapabilitiesEmitter.event;
+    }
+    get onDidChangeFile(): Event<readonly FileChange[]> {
+        return this.onDidChangeFileEmitter.event;
+    }
+    get onFileWatchError(): Event<void> {
+        return this.onFileWatchErrorEmitter.event;
+    }
+    async readFile(resource: URI): Promise<Uint8Array> {
+        const path = this.getHostPath(resource);
+        if (this.yjs.share.has(path)) {
+            const stringValue = this.yjs.getText(path);
+            return this.encoder.encode(stringValue.toString());
```

Comment:
It's not great what we do here, but we don't get the string buffer directly from yjs, so we need to encode it to an array buffer to satisfy the API. I don't think this will lead to actual performance issues though.